### PR TITLE
Move selectedChat logic into DeltachatController

### PR DESCRIPTION
**Pull request by Jikstra**
Currently wip, how can i force a rerender from inside selectChat()? ^^

### CODE REVIEW

File: src/main/deltachat.js [hunk 1]
```diff
@@ -278,26 +279,31 @@ class DeltaChatController {
   /**
    * Dispatched from menu alternative in SplittedChatListAndView
    */
-  leaveGroup (chatId) {
+  leaveGroup (chatId) {selectChat
```
Comment (ralphtheninja) [inline src/main/deltachat.js]: woops

File: src/main/deltachat.js [hunk 2]
```diff
@@ -278,26 +279,31 @@ class DeltaChatController {
...
     this._dc.removeContactFromChat(chatId, C.DC_CONTACT_ID_SELF)
   }
 
-  markNoticedChat (chatId) {
-    log('mark noticed chat', chatId)
-    this._dc.markNoticedChat(chatId)
+  selectChat(chatId) {
+    log('selecting chat with id', chatId)
+    this._selectedChatId = chatId
+    electron.ipcMain.emit('render')
```
Comment (ralphtheninja) [inline src/main/deltachat.js]: `this._render()`

File: src/main/deltachat.js [hunk 3]
```diff
@@ -278,26 +279,31 @@ class DeltaChatController {
...
   render () {
+    let chats = this._chats()
+
     return {
       configuring: this.configuring,
       credentials: this.credentials,
       ready: this.ready,
-      chats: this._chats(),
-      contacts: this._contacts()
+      chats: chats,
```
Comment (ralphtheninja) [inline src/main/deltachat.js]: Can do just `chats` here.

Comment (karissa) [reply src/main/deltachat.js]: or `chats: this._chats()` which would match the below contacts()

File: src/main/deltachat.js [hunk 4]
```diff
@@ -285,25 +285,25 @@ class DeltaChatController {
...
-    log(electron)
+    this._render()
   }
 
   /**
    * Returns the state in json format
    */
   render () {
     let chats = this._chats()
+    console.log('render')
```
Comment (ralphtheninja) [inline src/main/deltachat.js]: I added this before:

https://github.com/deltachat/deltachat-desktop/blob/40bed31b8d12b188b94e1d6b2aeff00b66b45c41/src/main/ipc.js#L94

File: src/main/deltachat.js [hunk 5]
```diff
@@ -319,6 +325,24 @@ class DeltaChatController {
     return chats
   }
 
+  selectedChat (chats) {
+    if (!chats) return null
+
+    let selectedChat = chats.find(({ id }) => id === this._selectedChatId)
```
Comment (ralphtheninja) [inline src/main/deltachat.js]: nice!

File: src/main/deltachat.js [hunk 6]
```diff
@@ -319,6 +324,24 @@ class DeltaChatController {
...
+
+    if (!selectedChat) {
+      selectedChat = chats.find(chat => chat.id !== C.DC_CHAT_ID_ARCHIVED_LINK)
```
Comment (ralphtheninja) [inline src/main/deltachat.js]: Use destructuring here as well!

File: src/renderer/components/SplittedChatListAndView.js [hunk 7]
```diff
@@ -138,27 +123,14 @@ class SplittedChatListAndView extends React.Component {
...
-    let { selectedChatId } = this.state
-
-    if (!selectedChatId) {
-      selectedChatId = this.state.selectedChatId = this.getInitiallySelectedChatId()
-    }
 
-    // TODO: We shouldn't do this on a render, maybe before? Maybe even in deltachat.js?
-    this.markNoticedChatIfNeeded(selectedChatId)
+    const selectedChat = deltachat.selectedChat
+    const selectedChatId = selectedChat.id
```
[the remark this answers is not in the document]
Comment (Jikstra) [reply src/renderer/components/SplittedChatListAndView.js]: nope, thanks :)

File: src/renderer/components/SplittedChatListAndView.js [hunk 8]
```diff
@@ -41,45 +40,26 @@ class SplittedChatListAndView extends React.Component {
...
-    const chatId = this.state.selectedChatId
-    this.state.selectedChatId = null
-    ipcRenderer.send('dispatch', 'archiveChat', chatId)
+  onArchiveChat (selectedChat) {
+    ipcRenderer.send('dispatch', 'archiveChat', selectedChat.id)
   }
 
-  onDeleteChat () {
-    const chatId = this.state.selectedChatId
-    this.state.selectedChatId = null
```
Comment (ralphtheninja) [inline src/renderer/components/SplittedChatListAndView.js]: I think this breaks select

File: src/renderer/components/SplittedChatListAndView.js [hunk 9]
```diff
@@ -41,45 +40,26 @@ class SplittedChatListAndView extends React.Component {
...
 
   onChatClick (chatId) {
-    this.markNoticedChatIfNeeded(chatId)
-    this.setState({ selectedChatId: chatId })
+    ipcRenderer.send('dispatch', 'selectChat', chatId)
   }
 
-  onArchiveChat () {
-    const chatId = this.state.selectedChatId
-    this.state.selectedChatId = null
```
Comment (ralphtheninja) [inline src/renderer/components/SplittedChatListAndView.js]: This break as well I think

**Pull request closed without merging**